File: compatibility-tests/sdk-test-node/tests/s3tables.test.ts
import { afterAll, beforeAll, describe, expect, it } from 'vitest';
import {
  ConflictException,
  CreateNamespaceCommand,
  CreateTableBucketCommand,
  CreateTableCommand,
  DeleteNamespaceCommand,
  DeleteTableBucketCommand,
  DeleteTableBucketPolicyCommand,
  DeleteTableCommand,
  DeleteTablePolicyCommand,
  GetNamespaceCommand,
  GetTableBucketCommand,
  GetTableBucketMaintenanceConfigurationCommand,
  GetTableBucketPolicyCommand,
  GetTableCommand,
  GetTableMaintenanceConfigurationCommand,
  GetTableMetadataLocationCommand,
  GetTablePolicyCommand,
  ListNamespacesCommand,
  ListTableBucketsCommand,
  ListTablesCommand,
  NotFoundException,
  PutTableBucketMaintenanceConfigurationCommand,
  PutTableBucketPolicyCommand,
  PutTableMaintenanceConfigurationCommand,
  PutTablePolicyCommand,
  RenameTableCommand,
  S3TablesClient,
  UpdateTableMetadataLocationCommand,
} from '@aws-sdk/client-s3tables';
import { makeClient, uniqueName } from './setup';

describe('S3 Tables lifecycle', () => {
  let client: S3TablesClient;
  let bucketArn: string;
  const bucketName = uniqueName('table-bucket');
  const namespace = 'analytics';
  let tableName = uniqueName('events');
  let versionToken: string;

  beforeAll(() => {
    client = makeClient(S3TablesClient);
  });

  afterAll(async () => {
    if (!bucketArn) return;
    await client.send(new DeleteTableCommand({ tableBucketARN: bucketArn, namespace, name: tableName })).catch(() => undefined);
    await client.send(new DeleteNamespaceCommand({ tableBucketARN: bucketArn, namespace })).catch(() => undefined);
    await client.send(new DeleteTableBucketCommand({ tableBucketARN: bucketArn })).catch(() => undefined);
  });

  it('creates a table bucket, namespace, and Iceberg table', async () => {
    const bucket = await client.send(new CreateTableBucketCommand({
      name: bucketName,
      tags: { suite: 'sdk-node' },
      encryptionConfiguration: { sseAlgorithm: 'AES256' },
      storageClassConfiguration: { storageClass: 'STANDARD' },
    }));
    bucketArn = bucket.arn!;
    expect(bucketArn).toContain(`bucket/${bucketName}`);

    const createdNamespace = await client.send(new CreateNamespaceCommand({
      tableBucketARN: bucketArn,
      namespace: [namespace],
    }));
    expect(createdNamespace.namespace).toEqual([namespace]);

    const table = await client.send(new CreateTableCommand({
      tableBucketARN: bucketArn,
      namespace,
      name: tableName,
      format: 'ICEBERG',
      metadata: { iceberg: { properties: { 'format-version': '2' } } },
      encryptionConfiguration: { sseAlgorithm: 'AES256' },
      storageClassConfiguration: { storageClass: 'STANDARD' },
      tags: { suite: 'sdk-node' },
    }));
    versionToken = table.versionToken!;
    expect(table.tableARN).toBe(`${bucketArn}/table/${tableName}`);
  });

  it('gets resources and applies list prefixes', async () => {
    const bucket = await client.send(new GetTableBucketCommand({ tableBucketARN: bucketArn }));
    expect(bucket.name).toBe(bucketName);

    const buckets = await client.send(new ListTableBucketsCommand({ prefix: 'table-bucket' }));
    expect(buckets.tableBuckets?.some((item) => item.name === bucketName)).toBe(true);

    const fetchedNamespace = await client.send(new GetNamespaceCommand({ tableBucketARN: bucketArn, namespace }));
    expect(fetchedNamespace.namespace).toEqual([namespace]);
    const namespaces = await client.send(new ListNamespacesCommand({ tableBucketARN: bucketArn, prefix: 'anal' }));
    expect(namespaces.namespaces?.some((item) => item.namespace?.includes(namespace))).toBe(true);

    const tables = await client.send(new ListTablesCommand({ tableBucketARN: bucketArn, namespace, prefix: 'events' }));
    expect(tables.tables?.some((table) => table.name === tableName)).toBe(true);
    const table = await client.send(new GetTableCommand({ tableBucketARN: bucketArn, namespace, name: tableName }));
    expect(table.tableARN).toBe(`${bucketArn}/table/${tableName}`);
  });

  it('round-trips policies and metadata version tokens', async () => {
    const policy = '{"Version":"2012-10-17","Statement":[]}';
    await client.send(new PutTableBucketPolicyCommand({ tableBucketARN: bucketArn, resourcePolicy: policy }));
    expect((await client.send(new GetTableBucketPolicyCommand({ tableBucketARN: bucketArn }))).resourcePolicy).toBe(policy);
    await client.send(new PutTablePolicyCommand({ tableBucketARN: bucketArn, namespace, name: tableName, resourcePolicy: policy }));
    expect((await client.send(new GetTablePolicyCommand({ tableBucketARN: bucketArn, namespace, name: tableName }))).resourcePolicy).toBe(policy);

    const updated = await client.send(new UpdateTableMetadataLocationCommand({
      tableBucketARN: bucketArn,
      namespace,
      name: tableName,
      metadataLocation: `s3://warehouse/${tableName}/v2.metadata.json`,
      versionToken,
    }));
    expect(updated.versionToken).not.toBe(versionToken);
    expect((await client.send(new GetTableMetadataLocationCommand({
      tableBucketARN: bucketArn, namespace, name: tableName,
    }))).metadataLocation).toBe(updated.metadataLocation);

    await expect(client.send(new UpdateTableMetadataLocationCommand({
      tableBucketARN: bucketArn,
      namespace,
      name: tableName,
      metadataLocation: `s3://warehouse/${tableName}/stale.metadata.json`,
      versionToken,
    }))).rejects.toBeInstanceOf(ConflictException);
    versionToken = updated.versionToken!;
  });

  it('round-trips typed maintenance configuration maps', async () => {
    await client.send(new PutTableBucketMaintenanceConfigurationCommand({
      tableBucketARN: bucketArn,
      type: 'icebergUnreferencedFileRemoval',
      value: { status: 'enabled' },
    }));
    const bucketMaintenance = await client.send(new GetTableBucketMaintenanceConfigurationCommand({
      tableBucketARN: bucketArn,
    }));
    expect(bucketMaintenance.configuration?.icebergUnreferencedFileRemoval?.status).toBe('enabled');

    await client.send(new PutTableMaintenanceConfigurationCommand({
      tableBucketARN: bucketArn,
      namespace,
      name: tableName,
      type: 'icebergCompaction',
      value: { status: 'enabled' },
    }));
    const tableMaintenance = await client.send(new GetTableMaintenanceConfigurationCommand({
      tableBucketARN: bucketArn,
      namespace,
      name: tableName,
    }));
    expect(tableMaintenance.configuration?.icebergCompaction?.status).toBe('enabled');
  });

  it('renames a table and rotates its version token', async () => {
    const originalToken = versionToken;
    const renamedTableName = `${tableName}-renamed`;
    await client.send(new RenameTableCommand({
      tableBucketARN: bucketArn,
      namespace,
      name: tableName,
      newName: renamedTableName,
      versionToken: originalToken,
    }));

    const renamed = await client.send(new GetTableCommand({
      tableBucketARN: bucketArn,
      namespace,
      name: renamedTableName,
    }));
    expect(renamed.name).toBe(renamedTableName);
    expect(renamed.versionToken).not.toBe(originalToken);
    await expect(client.send(new GetTableCommand({ tableBucketARN: bucketArn, namespace, name: tableName })))
      .rejects.toBeInstanceOf(NotFoundException);
    tableName = renamedTableName;
    versionToken = renamed.versionToken!;
  });

  it('deletes policies and the resource hierarchy', async () => {
    await client.send(new DeleteTablePolicyCommand({ tableBucketARN: bucketArn, namespace, name: tableName }));
    await expect(client.send(new GetTablePolicyCommand({ tableBucketARN: bucketArn, namespace, name: tableName })))
      .rejects.toBeInstanceOf(NotFoundException);
    await client.send(new DeleteTableBucketPolicyCommand({ tableBucketARN: bucketArn }));
    await expect(client.send(new GetTableBucketPolicyCommand({ tableBucketARN: bucketArn })))
      .rejects.toBeInstanceOf(NotFoundException);

    await client.send(new DeleteTableCommand({ tableBucketARN: bucketArn, namespace, name: tableName }));
    await expect(client.send(new GetTableCommand({ tableBucketARN: bucketArn, namespace, name: tableName })))
      .rejects.toBeInstanceOf(NotFoundException);
    await client.send(new DeleteNamespaceCommand({ tableBucketARN: bucketArn, namespace }));
    await expect(client.send(new GetNamespaceCommand({ tableBucketARN: bucketArn, namespace })))
      .rejects.toBeInstanceOf(NotFoundException);
    await client.send(new DeleteTableBucketCommand({ tableBucketARN: bucketArn }));
    await expect(client.send(new GetTableBucketCommand({ tableBucketARN: bucketArn })))
      .rejects.toBeInstanceOf(NotFoundException);
  });
});

File: compatibility-tests/sdk-test-java/src/test/java/com/floci/test/S3TablesTest.java
package com.floci.test;

import org.junit.jupiter.api.AfterAll;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.MethodOrderer;
import org.junit.jupiter.api.Order;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.TestMethodOrder;
import software.amazon.awssdk.services.s3tables.S3TablesClient;
import software.amazon.awssdk.services.s3tables.model.*;

import java.util.Map;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

@TestMethodOrder(MethodOrderer.OrderAnnotation.class)
class S3TablesTest {
    private static S3TablesClient client;
    private static String bucketName;
    private static String bucketArn;
    private static String tableName;
    private static String versionToken;
    private static final String NAMESPACE = "analytics";

    @BeforeAll
    static void setup() {
        client = TestFixtures.s3tablesClient();
        bucketName = TestFixtures.uniqueName("table-bucket");
        tableName = TestFixtures.uniqueName("events");
    }

    @AfterAll
    static void cleanup() {
        if (client == null) {
            return;
        }
        if (bucketArn != null) {
            cleanup("table " + tableName, () -> client.deleteTable(DeleteTableRequest.builder()
                    .tableBucketARN(bucketArn).namespace(NAMESPACE).name(tableName).build()));
            cleanup("namespace " + NAMESPACE, () -> client.deleteNamespace(DeleteNamespaceRequest.builder()
                    .tableBucketARN(bucketArn).namespace(NAMESPACE).build()));
            cleanup("table bucket " + bucketArn, () -> client.deleteTableBucket(DeleteTableBucketRequest.builder()
                    .tableBucketARN(bucketArn).build()));
        }
        client.close();
    }

    private static void cleanup(String resource, Runnable deletion) {
        try {
            deletion.run();
        } catch (Exception exception) {
            System.err.printf("S3 Tables SDK test cleanup failed for %s: %s%n", resource, exception.getMessage());
        }
    }

    @Test
    @Order(1)
    void createsAndGetsTableBucket() {
        CreateTableBucketResponse created = client.createTableBucket(
                CreateTableBucketRequest.builder()
                        .name(bucketName)
                        .tags(Map.of("suite", "sdk-java"))
                        .encryptionConfiguration(EncryptionConfiguration.builder().sseAlgorithm(SSEAlgorithm.AES256).build())
                        .storageClassConfiguration(StorageClassConfiguration.builder().storageClass(StorageClass.STANDARD).build())
                        .build());
        bucketArn = created.arn();
        assertThat(bucketArn).contains("arn:aws:s3tables:").endsWith("bucket/" + bucketName);

        GetTableBucketResponse fetched = client.getTableBucket(
                GetTableBucketRequest.builder().tableBucketARN(bucketArn).build());
        assertThat(fetched.name()).isEqualTo(bucketName);
        assertThat(fetched.arn()).isEqualTo(bucketArn);
    }

    @Test
    @Order(2)
    void createsNamespaceAndTable() {
        CreateNamespaceResponse namespace = client.createNamespace(
                CreateNamespaceRequest.builder().tableBucketARN(bucketArn).namespace(NAMESPACE).build());
        assertThat(namespace.namespace()).containsExactly(NAMESPACE);

        CreateTableResponse table = client.createTable(CreateTableRequest.builder()
                .tableBucketARN(bucketArn)
                .namespace(NAMESPACE)
                .name(tableName)
                .format(OpenTableFormat.ICEBERG)
                .metadata(TableMetadata.builder().iceberg(IcebergMetadata.builder()
                        .properties(Map.of("format-version", "2")).build()).build())
                .encryptionConfiguration(EncryptionConfiguration.builder().sseAlgorithm(SSEAlgorithm.AES256).build())
                .storageClassConfiguration(StorageClassConfiguration.builder().storageClass(StorageClass.STANDARD).build())
                .tags(Map.of("suite", "sdk-java"))
                .build());
        versionToken = table.versionToken();
        assertThat(table.tableARN()).isEqualTo(bucketArn + "/table/" + tableName);
        assertThat(versionToken).isNotBlank();

        GetTableResponse fetched = client.getTable(GetTableRequest.builder()
                .tableBucketARN(bucketArn).namespace(NAMESPACE).name(tableName).build());
        assertThat(fetched.tableARN()).isEqualTo(table.tableARN());
        assertThat(fetched.namespace()).containsExactly(NAMESPACE);
        assertThat(fetched.format()).isEqualTo(OpenTableFormat.ICEBERG);
    }

    @Test
    @Order(3)
    void getsAndListsResourcesWithPrefixes() {
        assertThat(client.listTableBuckets(ListTableBucketsRequest.builder().prefix("table-bucket").build()).tableBuckets())
                .anyMatch(bucket -> bucketName.equals(bucket.name()));
        assertThat(client.getNamespace(GetNamespaceRequest.builder().tableBucketARN(bucketArn).namespace(NAMESPACE).build())
                .namespace()).containsExactly(NAMESPACE);
        assertThat(client.listNamespaces(ListNamespacesRequest.builder().tableBucketARN(bucketArn).prefix("anal").build())
                .namespaces()).anyMatch(namespace -> namespace.namespace().contains(NAMESPACE));
        assertThat(client.listTables(ListTablesRequest.builder().tableBucketARN(bucketArn).namespace(NAMESPACE)
                        .prefix("events").build()).tables())
                .anyMatch(table -> tableName.equals(table.name()));
    }

    @Test
    @Order(4)
    void updatesAndGetsMetadataLocationAndRejectsStaleVersionToken() {
        UpdateTableMetadataLocationResponse updated = client.updateTableMetadataLocation(
                UpdateTableMetadataLocationRequest.builder()
                        .tableBucketARN(bucketArn)
                        .namespace(NAMESPACE)
                        .name(tableName)
                        .metadataLocation("s3://warehouse/" + tableName + "/v2.metadata.json")
                        .versionToken(versionToken)
                        .build());
        assertThat(updated.metadataLocation()).endsWith("v2.metadata.json");
        assertThat(updated.versionToken()).isNotEqualTo(versionToken);
        assertThat(client.getTableMetadataLocation(GetTableMetadataLocationRequest.builder()
                        .tableBucketARN(bucketArn).namespace(NAMESPACE).name(tableName).build())
                .metadataLocation()).isEqualTo(updated.metadataLocation());

        assertThatThrownBy(() -> client.updateTableMetadataLocation(
                UpdateTableMetadataLocationRequest.builder()
                        .tableBucketARN(bucketArn).namespace(NAMESPACE).name(tableName)
                        .metadataLocation("s3://warehouse/stale.metadata.json")
                        .versionToken(versionToken).build()))
                .isInstanceOf(ConflictException.class);
        versionToken = updated.versionToken();
    }

    @Test
    @Order(5)
    void roundTripsAndDeletesBucketAndTablePolicies() {
        String policy = "{\"Version\":\"2012-10-17\",\"Statement\":[]}";
        client.putTableBucketPolicy(PutTableBucketPolicyRequest.builder()
                .tableBucketARN(bucketArn).resourcePolicy(policy).build());
        assertThat(client.getTableBucketPolicy(GetTableBucketPolicyRequest.builder().tableBucketARN(bucketArn).build())
                .resourcePolicy()).isEqualTo(policy);

        client.putTablePolicy(PutTablePolicyRequest.builder()
                .tableBucketARN(bucketArn).namespace(NAMESPACE).name(tableName).resourcePolicy(policy).build());
        assertThat(client.getTablePolicy(GetTablePolicyRequest.builder()
                .tableBucketARN(bucketArn).namespace(NAMESPACE).name(tableName).build()).resourcePolicy()).isEqualTo(policy);
    }

    @Test
    @Order(6)
    void roundTripsTypedMaintenanceConfigurations() {
        TableBucketMaintenanceConfigurationValue bucketValue = TableBucketMaintenanceConfigurationValue.builder()
                .status(MaintenanceStatus.ENABLED).build();
        client.putTableBucketMaintenanceConfiguration(PutTableBucketMaintenanceConfigurationRequest.builder()
                .tableBucketARN(bucketArn)
                .type(TableBucketMaintenanceType.ICEBERG_UNREFERENCED_FILE_REMOVAL)
                .value(bucketValue)
                .build());
        assertThat(client.getTableBucketMaintenanceConfiguration(GetTableBucketMaintenanceConfigurationRequest.builder()
                        .tableBucketARN(bucketArn).build())
                .configuration().get(TableBucketMaintenanceType.ICEBERG_UNREFERENCED_FILE_REMOVAL).status())
                .isEqualTo(MaintenanceStatus.ENABLED);

        TableMaintenanceConfigurationValue tableValue = TableMaintenanceConfigurationValue.builder()
                .status(MaintenanceStatus.ENABLED).build();
        client.putTableMaintenanceConfiguration(PutTableMaintenanceConfigurationRequest.builder()
                .tableBucketARN(bucketArn).namespace(NAMESPACE).name(tableName)
                .type(TableMaintenanceType.ICEBERG_COMPACTION)
                .value(tableValue)
                .build());
        assertThat(client.getTableMaintenanceConfiguration(GetTableMaintenanceConfigurationRequest.builder()
                        .tableBucketARN(bucketArn).namespace(NAMESPACE).name(tableName).build())
                .configuration().get(TableMaintenanceType.ICEBERG_COMPACTION).status())
                .isEqualTo(MaintenanceStatus.ENABLED);
    }

    @Test
    @Order(7)
    void renamesTableAndRotatesItsVersionToken() {
        String originalToken = versionToken;
        String renamedTableName = tableName + "-renamed";
        client.renameTable(RenameTableRequest.builder()
                .tableBucketARN(bucketArn).namespace(NAMESPACE).name(tableName)
                .newName(renamedTableName).versionToken(originalToken).build());

        GetTableResponse renamed = client.getTable(GetTableRequest.builder()
                .tableBucketARN(bucketArn).namespace(NAMESPACE).name(renamedTableName).build());
        assertThat(renamed.name()).isEqualTo(renamedTableName);
        assertThat(renamed.versionToken()).isNotEqualTo(originalToken);
        assertThatThrownBy(() -> client.getTable(GetTableRequest.builder()
                .tableBucketARN(bucketArn).namespace(NAMESPACE).name(tableName).build()))
                .isInstanceOf(NotFoundException.class);
        tableName = renamedTableName;
        versionToken = renamed.versionToken();
    }

    @Test
    @Order(8)
    void deletesPoliciesAndResourceHierarchy() {
        client.deleteTablePolicy(DeleteTablePolicyRequest.builder()
                .tableBucketARN(bucketArn).namespace(NAMESPACE).name(tableName).build());
        assertThatThrownBy(() -> client.getTablePolicy(GetTablePolicyRequest.builder()
                .tableBucketARN(bucketArn).namespace(NAMESPACE).name(tableName).build()))
                .isInstanceOf(NotFoundException.class);

        client.deleteTableBucketPolicy(DeleteTableBucketPolicyRequest.builder().tableBucketARN(bucketArn).build());
        assertThatThrownBy(() -> client.getTableBucketPolicy(GetTableBucketPolicyRequest.builder().tableBucketARN(bucketArn).build()))
                .isInstanceOf(NotFoundException.class);

        client.deleteTable(DeleteTableRequest.builder().tableBucketARN(bucketArn).namespace(NAMESPACE).name(tableName).build());
        assertThatThrownBy(() -> client.getTable(GetTableRequest.builder()
                .tableBucketARN(bucketArn).namespace(NAMESPACE).name(tableName).build()))
                .isInstanceOf(NotFoundException.class);
        client.deleteNamespace(DeleteNamespaceRequest.builder().tableBucketARN(bucketArn).namespace(NAMESPACE).build());
        assertThatThrownBy(() -> client.getNamespace(GetNamespaceRequest.builder()
                .tableBucketARN(bucketArn).namespace(NAMESPACE).build()))
                .isInstanceOf(NotFoundException.class);
        client.deleteTableBucket(DeleteTableBucketRequest.builder().tableBucketARN(bucketArn).build());
        assertThatThrownBy(() -> client.getTableBucket(GetTableBucketRequest.builder().tableBucketARN(bucketArn).build()))
                .isInstanceOf(NotFoundException.class);
    }
}
